### PR TITLE
Migrations now use virtual methods and have a concept of a failed migration

### PR DESCRIPTION
**Pull request by ChaseFlorell**
There's always room for improvement, but this is a good first run.

A couple of questions come to mind however, and I'm not sure the right answer.

 1. if a migration fails, should it halt all subsequent migrations?
 2. should client/server migrations run BEFORE shared migrations or after? <sup>*</sup>
 3. When using `DateTimeOffset`, should I be calling `.UtcNow` or `.Now`?

> the migration runner has the ability to either separate out a migration based on Server / Client, or to run the Migration on both platforms. By having your migration run on both, you'll get the benefit of a matching schema.

### CODE REVIEW

File: src/crossql/Migrations/MigrationRunner.cs [hunk 1]
```diff
@@ -20,182 +19,230 @@ public MigrationRunner(IDbProvider dbProvider)
...
             {
-                databaseVersion = new DatabaseVersionModel
-                {
-                    VersionNumber = migration.MigrationVersion,
-                    MigrationDate = DateTime.UtcNow
-                };
-                await _dbProvider.Create(databaseVersion).ConfigureAwait(false);
-            }
+                VersionNumber = migration.MigrationVersion,
+                MigrationDate = DateTimeOffset.Now
```
Comment (ChaseFlorell) [inline src/crossql/Migrations/MigrationRunner.cs:221]: Should we be using `DateTimeOffset.Now` or `DateTimeOffset.UtcNow`?

File: src/crossql/Migrations/MigrationRunner.cs [hunk 2]
```diff
@@ -20,182 +19,230 @@ public MigrationRunner(IDbProvider dbProvider)
...
+                }
 
-        private async Task<DatabaseVersionModel> GetMigrationInformationAsync(IDbMigration migration)
+                var dbCommand = database.ToString();
+                if (!string.IsNullOrWhiteSpace(dbCommand))
+                    await _dbProvider.ExecuteNonQuery(dbCommand);
+            }
+            catch (Exception ex)
+            {
+                var database = new Database(dbProvider.DatabaseName, dbProvider.Dialect);
```
Comment (ChaseFlorell) [inline src/crossql/Migrations/MigrationRunner.cs:202]: If any one of the migrations fail, we run the failed migration method and exit. Is this the best way, or should subsequent migrations still continue?

_I feel as though they should not_

File: src/crossql/Migrations/MigrationRunner.cs [hunk 3]
```diff
@@ -20,182 +19,230 @@ public MigrationRunner(IDbProvider dbProvider)
...
-
-                if (_systemRole == SystemRole.Server)
+                switch (systemRole)
                 {
-                    await migration.RunOrderedMigration(MigrationStep.ServerSetup, _dbProvider).ConfigureAwait(false);
-                }
-                if (_systemRole == SystemRole.Client)
-                {
-                    await migration.RunOrderedMigration(MigrationStep.ClientSetup, _dbProvider).ConfigureAwait(false);
+                    case SystemRole.Server:
```
Comment (ChaseFlorell) [inline src/crossql/Migrations/MigrationRunner.cs:66]: right now it's setup that system specific migrations run BEFORE shared migrations. 
In a perfect world, migrations would be idempotent, but it's just a question... is this the best approach?

File: src/crossql/Migrations/MigrationRunner.cs (hunk 1 above)
Comment (ChaseFlorell) [reply src/crossql/Migrations/MigrationRunner.cs:221]: note: this really doesn't matter all that much since it's not a table that's designed to sync.

**Comment by dylanberry** [comment]
> 1. if a migration fails, should it halt all subsequent migrations?
> 2. should client/server migrations run BEFORE shared migrations or after?
> 3. When using DateTimeOffset, should I be calling .UtcNow or .Now?

1. Yes, for a KISS approach, migrations should be treated as a single system level transaction, each must proceed before moving to the next. We could build something around migration dependencies which would allow some to proceed without others, that would of course add complexity.
2. What do you mean by client/server migrations and shared migrations?
3. `.Now()` for display. IMO, if you're not displaying it, keep it as `DateTimeOffset`.

**Comment by ChaseFlorell** [comment]
> What do you mean by client/server migrations and shared migrations?

The system is designed to allow you to have a Migration step that runs on Client || Server and also a Migration step that runs on Client && Server

```
Task Migrate(Database database, IDbProvider provider); // runs on the client and server
Task ClientMigrate(Database database, IDbProvider provider); // runs only on the client
Task ServerMigrate(Database database, IDbProvider provider); // runs only on the server
```

This might have to live in the documentation

 > If you choose to use `ClientMigrate` / `ServerMigrate` in conjunction with `Migrate`, just know that the platform specific migration will run before the shared migration. It is best that if one migration depends on another that you take advantage of `SetupMigration`, `[Platform]SetupMigration`, and `[Platform]FinishMigration`. If any migration step fails, you can override MigrationFailed where you will have access to which MigrationStep failed, and can rollback accordingly.

**Comment by ChaseFlorell** [comment]
> .Now() for display. IMO, if you're not displaying it, keep it as DateTimeOffset.

right now we're not displaying anything, this is just a record of when a specific migration was run. I actually think it's posterity over anything else.

**Comment by ChaseFlorell** [comment]
> Yes, for a KISS approach, migrations should be treated as a single system level transaction, each must proceed before moving to the next. We could build something around migration dependencies which would allow some to proceed without others, that would of course add complexity.

The way I've written it, if you have 3 migrations and Migration002 fails, Migration003 will not run.

This test demonstrates
https://github.com/crossql/crossql/pull/23/files#diff-ee06db5cf532b118bcc39217bfd4e28fR54

**Comment by ChaseFlorell** [comment]
As far as question (3), I'm happy with `DateTimeOffset.Now`.
https://stackoverflow.com/a/14268167/124069

**Pull request merged**